tensorflow_graphics/projects/mesh_rcnn/ops/cubify.py

```diff
@@ -0,0 +1,303 @@
+
+  batch_sizes = voxel_grid.shape[:-3].as_list()
+  if len(batch_sizes) == 0:
+    batch_sizes = None
+  # N x D x H x W.
+  voxel_grid = tf.reshape(voxel_grid, [-1] + voxel_grid.shape[-3:].as_list())
+
+  batch_size, depth, height, width = voxel_grid.shape
+  unit_cube_verts = tf.constant(
+      [
+          [0, 0, 0],
+          [0, 0, 1],
+          [0, 1, 0],
+          [0, 1, 1],
+          [1, 0, 0],
+          [1, 0, 1],
+          [1, 1, 0],
+          [1, 1, 1],
+      ], dtype=tf.int32)
+
+  unit_cube_faces = tf.constant(
+      [
+          [0, 1, 2],
+          [1, 3, 2],  # left face: 0, 1
+          [2, 3, 6],
+          [3, 7, 6],  # bottom face: 2, 3
+          [0, 2, 6],
+          [0, 6, 4],  # front face: 4, 5
+          [0, 5, 1],
+          [0, 4, 5],  # up face: 6, 7
+          [6, 7, 5],
+          [6, 5, 4],  # right face: 8, 9
+          [1, 7, 3],
+          [1, 5, 7],  # back face: 10, 11
+      ],
+      dtype=tf.int32
+  )
+
+  voxel_mask = tf.math.greater_equal(voxel_grid, threshold)
+  voxel_thresholded = tf.cast(voxel_mask, tf.float32)
+
+  if tf.reduce_all(tf.math.logical_not(voxel_mask)):
+    return Meshes([], [])
+
+  kernel_x = tf.constant([.5, .5], shape=(1, 1, 2, 1, 1), dtype=tf.float32)
+  kernel_y = tf.constant([.5, .5], shape=(1, 2, 1, 1, 1), dtype=tf.float32)
+  kernel_z = tf.constant([.5, .5], shape=(2, 1, 1, 1, 1), dtype=tf.float32)
+
+  x_left_faces, x_right_faces = _create_face_mask(voxel_thresholded, kernel_x,
+                                                  -1)
+  y_top_faces, y_bottom_faces = _create_face_mask(voxel_thresholded, kernel_y,
+                                                  -2)
+  z_front_faces, z_back_faces = _create_face_mask(voxel_thresholded, kernel_z,
+                                                  -3)
+
+  faces_indices = [
+      x_left_faces,
+      x_left_faces,  # faces 0, 1
+      y_bottom_faces,
+      y_bottom_faces,  # faces 2, 3
+      z_front_faces,
+      z_front_faces,  # faces 4, 5
+      y_top_faces,
+      y_top_faces,  # faces 6, 7
+      x_right_faces,
+      x_right_faces,  # faces 8, 9
+      z_back_faces,
+      z_back_faces,  # faces 10, 11
+  ]
+
+  # 12 x batch_size x depth x height x width
+  faces_idx = tf.stack(faces_indices, axis=0)
+
+  # batch_size x height x width x depth x 12
+  faces_idx = tf.transpose(faces_idx, perm=(1, 3, 4, 2, 0))
+
+  faces_idx = tf.reshape(faces_idx, shape=(-1, unit_cube_faces.shape[0]))
+  linear_index = tf.cast(tf.where(tf.not_equal(faces_idx, 0)), tf.int32)
+  nyxz = tf.transpose(tf.unravel_index(linear_index[:, 0],
+                                       (batch_size, height, width, depth)))
+
+  if len(nyxz) == 0:
+    return Meshes([], [])
+
+  unit_cube_faces_per_occupied_voxel = tf.gather(unit_cube_faces,
+                                                 linear_index[:, 1],
+                                                 axis=None)
+
+  def _compute_grid_coordinates(face_index):
```

Comment:
Move this helper method outside the function definition, or put it on top. 
Add a single line summary for the helper.

Reply:
I would leave it inside cubify, because we need acces to the cubify scope (e.g. `unit_cube_verts`, `height`, `width`, `depth`, ...). Otherwise we would have to pass those as arguments in `tf.vectorized_map`.